lib/rubocop/cop/rspec/subject_stub.rb

```diff
@@ -74,71 +75,59 @@ class SubjectStub < Cop
         PATTERN
 
         def on_block(node)
+          # process only describe/context/... example groups
           return unless example_group?(node)
 
-          find_subject_stub(node) do |stub|
+          # skip already processed example group
+          # it's processed if is nested in one of the processed example groups
+          return unless (processed_example_groups & node.ancestors).empty?
```

Comment:
How big are your specs so this makes a significant difference?

Reply:
I ran Rubocop on its own source code so the test suite isn't large. I believe the difference would be more significant on larger test code base.

Reply:
Regarding the performance testing in the wild - you can pick some projects out of [this list](https://github.com/eliotsykes/real-world-rails), I find https://github.com/discourse/discourse/ specifically good.

Reply:
Checked this optimization on Gitlab source code and got the following results:
- it speeds up the cope from 13%  to 5.8% of samples
- the whole execution time decreased by ~10s

Reply:
You can use `on_top_level describe` and then search for example groups, in order to avoid double lookups. E.g. once you are in an example group, you can break from going to nested example groups,

Reply:
@Darhazer This is a truly amazing proposal.

Reply:
But then it turns out that RuboCop::RSpec::TopLevelDescribe#on_send which triggers on_top_level_descrive, is one of [the slowest methods](https://github.com/rubocop-hq/rubocop/issues/8022)

Reply:
Ha, fair enough. Never looked closely at `TopLevelDescribe`. Frankly, I'm having a hard time understanding why it's implemented that way. Should be: "call `on_top_level_describe` if method name is `:describe`, and there are no ancestors of the type `:block` with a method name `:describe`". It'd better keep a set of already detected TLDs to quickly return from `on_send` if `(node.ancestors & top_level_describes).any?` just like Andrew has implemented in this pull request.

Luckily, it doesn't fit our purposes in this pull request anyway, since it's only for `describe`:
```ruby
return false unless node.method_name == :describe
```
while we need something like `TopLevelExampleGroup`, probably implemented in a way I suggest above, but focusing on all example group methods, not only `:describe`.

Would you like to give this idea a shot in the scope of this pull request, or do you feel this improvement that you've done is good on its own already, @andrykonchin?

Reply:
I would love to optimize `TopLevelDescribe` but think it deserves a separate PR.

Currently I am working on optimizing cops from the main `rubocop` gem but will definitely look at `rubocop-rspec` after that. 

P.S.

> But then it turns out that RuboCop::RSpec::TopLevelDescribe#on_send which triggers on_top_level_descrive, is one of the slowest methods

Switching from manual caching to `#on_top_level_describe` callback decreased profiler samples for `SubjectStub` cop from 0.9% to 0% so it isn't as bad as you may think 😉.